### ISSUE TITLE
Extraction does not correctly handle shadowing of record fields

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#16677]
```
File "/tmp/testextraction0c1364.ml", line 596274, characters 52-64:
596274 |           opts.parsed_synthesize_options.inline } { package_name =
                                                             ^^^^^^^^^^^^
Error: Unbound module Coq__9
Hint: Did you mean Coq__1?
File "/tmp/tmpizjts5kg/Crypto/CLI.v", line 1023, characters 0-60:
Error:
Compilation of file /tmp/testextraction0c1364.ml failed with exit code 2
```
You can reproduce this error with https://github.com/JasonGross/fiat-crypto/tree/xxx-coq-bug-extraction-bad, `make src/CLI.vo SKIP_BEDROCK2=1`.  (I'm working on minimizing this, but don't hold out hope.  Extraction takes around 35 seconds so I expect the minimization process to run at least a week or two.)

The relevant generated block of code looks like:
```ocaml
        synthesize0 { Pipeline.coq_AbstractInterpretation_opts =
          opts.parsed_synthesize_options.abstract_interpretation_options;
          Pipeline.coq_RewriteConfiguration_opts =
          { Pipeline.low_level_rewriter_method =
          (low_level_rewriter_method opts.parsed_synthesize_options);
          Pipeline.only_signed = opts.parsed_synthesize_options.only_signed;
          Pipeline.unfold_value_barrier =
          opts.parsed_synthesize_options.unfold_value_barrier;
          Pipeline.relax_adc_sbb_return_carry_to_bitwidth =
          opts.parsed_synthesize_options.output_options.Coq7_Compilers.Options.relax_adc_sbb_return_carry_to_bitwidth_ };
          Pipeline.coq_Pipeline_opts =
          opts.parsed_synthesize_options.machine_wordsize }
          { Pipeline.output_language_api = output_language_api0;
          Pipeline.language_naming_conventions =
          opts.parsed_synthesize_options.language_naming_conventions;
          Pipeline.documentation_options =
          opts.parsed_synthesize_options.documentation_options;
          Pipeline.output_options =
          opts.parsed_synthesize_options.output_options;
          Pipeline.internal_static =
          opts.parsed_synthesize_options.internal_static; Pipeline.static =
          opts.parsed_synthesize_options.static; Pipeline.all_static =
          (all_static opts.parsed_synthesize_options); Pipeline.inline =
          opts.parsed_synthesize_options.inline } { package_name0 =
          opts.parsed_synthesize_options.internal_package_name; class_name0 =
          opts.parsed_synthesize_options.internal_class_name; inline0 =
          opts.parsed_synthesize_options.inline; Coq__11.inline_internal =
          opts.parsed_synthesize_options.inline_internal; no_select0 =
          opts.parsed_synthesize_options.no_select;
          Coq__12.use_mul_for_cmovznz =
          opts.parsed_synthesize_options.use_mul_for_cmovznz;
          Coq__13.emit_primitives =
          opts.parsed_synthesize_options.emit_primitives; should_split_mul0 =
          opts.parsed_synthesize_options.should_split_mul;
          should_split_multiret0 =
          opts.parsed_synthesize_options.should_split_multiret;
          Coq__14.assembly_hints_lines = opts.assembly_hints_lines;
          widen_carry0 = opts.parsed_synthesize_options.widen_carry;
          widen_bytes0 = opts.parsed_synthesize_options.widen_bytes;
          Coq__15.error_on_unused_assembly_functions =
          opts.parsed_synthesize_options.error_on_unused_assembly_functions;
          Coq__16.ignore_unique_asm_names =
          opts.parsed_synthesize_options.ignore_unique_asm_names;
          Coq__17.assembly_conventions =
          opts.parsed_synthesize_options.assembly_conventions }
          tight_bounds_multiplier n0 s c comment_header prefix1 requests)) }
```

Note in particular the block that looks like (after reflowing)
```ocaml
{
  package_name0 = opts.parsed_synthesize_options.internal_package_name;
  class_name0 = opts.parsed_synthesize_options.internal_class_name;
  inline0 = opts.parsed_synthesize_options.inline;
  Coq__11.inline_internal = opts.parsed_synthesize_options.inline_internal;
  no_select0 = opts.parsed_synthesize_options.no_select;
  Coq__12.use_mul_for_cmovznz = opts.parsed_synthesize_options.use_mul_for_cmovznz;
  Coq__13.emit_primitives = opts.parsed_synthesize_options.emit_primitives;
  should_split_mul0 = opts.parsed_synthesize_options.should_split_mul;
  should_split_multiret0 = opts.parsed_synthesize_options.should_split_multiret;
  Coq__14.assembly_hints_lines = opts.assembly_hints_lines;
  widen_carry0 = opts.parsed_synthesize_options.widen_carry;
  widen_bytes0 = opts.parsed_synthesize_options.widen_bytes;
  Coq__15.error_on_unused_assembly_functions = opts.parsed_synthesize_options.error_on_unused_assembly_functions;
  Coq__16.ignore_unique_asm_names = opts.parsed_synthesize_options.ignore_unique_asm_names;
  Coq__17.assembly_conventions = opts.parsed_synthesize_options.assembly_conventions;
}
```
There is no way this could ever be valid OCaml, AIUI, because record fields must all come from the same module.  So something is seriously wrong with the way extraction is shadowing record field names.

**Comment by JasonGross** [comment]
Note that Coq already handles this correctly in the case of
```coq
(* -*- coq-prog-args: ("-top" "foo") -*- *)
Require Import Coq.extraction.Extraction.
Definition foo (b : bool) := if b then false else true.
Module test.
  Definition foo (b : bool) := if b then true else true.
  Definition bar := match foo.foo true, test.foo true with
                    | true, true => 0
                    | false, true => 1
                    | true, false => 2
                    | false, false => 3
                    end.
End test.
Recursive Extraction test.bar.
(*

type bool =
| True
| False

type nat =
| O
| S of nat

module Coq__1 = struct
 (** val foo : bool -> bool **)let foo = function
                               | True -> False
                               | False -> True
end
include Coq__1

module Coq_test =
 struct
  (** val foo : bool -> bool **)

  let foo _ =
    True

  (** val bar : nat **)

  let bar =
    match Coq__1.foo True with
    | True -> (match foo True with
               | True -> O
               | False -> S (S O))
    | False -> (match foo True with
                | True -> S O
                | False -> S (S (S O)))
 end

*)
```

**Comment by JasonGross** [comment]
This might be illustrative:
```coq
(* -*- coq-prog-args: ("-top" "foo") -*- *)
Require Import Coq.extraction.Extraction.
Class Foo := { foo : bool ; bar : bool }.
Definition qux f g := (@bar f, @bar g).
Module test.
  Class Foo := { foo : bool ; bar : bool }.
  Definition qux f g := (@bar f, @bar g).
  Definition baz := (foo.qux, qux).
End test.
Definition quxx := (@test.foo, @test.baz).
Recursive Extraction quxx.
(*

type bool =
| True
| False

type ('a, 'b) prod =
| Pair of 'a * 'b

type foo = { foo0 : bool; bar : bool }

module Coq__1 = struct
 (** val qux : foo -> foo -> (bool, bool) prod **)let qux f g =
                                                    Pair (f.bar, g.bar)
end
include Coq__1

module Coq_test =
 struct
  type coq_Foo = { foo : bool; bar : bool }

  (** val foo : coq_Foo -> bool **)

  let foo foo1 =
    foo1.foo

  (** val bar : coq_Foo -> bool **)

  let bar foo1 =
    foo1.bar

  (** val qux : coq_Foo -> coq_Foo -> (bool, bool) prod **)

  let qux f g =
    Pair (f.bar, g.bar)

  (** val baz : (foo -> foo -> (bool, bool) prod, coq_Foo -> coq_Foo -> (bool, bool) prod) prod **)

  let baz =
    Pair (Coq__1.qux, qux)
 end

(** val quxx :
    (Coq_test.coq_Foo -> bool, (foo -> foo -> (bool, bool) prod, Coq_test.coq_Foo -> Coq_test.coq_Foo -> (bool, bool) prod) prod) prod **)

let quxx =
  Pair ((fun foo1 -> foo1.Coq_test.foo), Coq_test.baz)

 *)
```
It seems that Coq will only create the `Coq__` modules for true constants that are shadowed, and it will create a separate module for each such shadowed constant.  Maybe the code needs to be extended to handle shadowed record projections?

**Comment by SkySkimmer** [comment]
Duplicate of https://github.com/coq/coq/issues/12813

**Comment by herbelin** [comment]
Fixed in #17324.
